### PR TITLE
improve(runTransaction): Pad gas for multicall3 aggregate

### PR DESCRIPTION
**Pull request by pxrl**
This change implements three key things:
        
1) Capture the gas estimates from successful txn simulations and pass it
through to txn submission.
        
2) Modify TransactionClient() to check for a gasLimitMultiplier argument
If a multiplier was supplied, apply it to the gasLimit at that point,
before passing a transaction request over to runTransaction() for 
submission to the RPC provider.
        
3) Modify the MultiCallerClient class to apply 20% gas padding on
multicall3 aggregate() transactions. This addresses a specific issue
that has been seen with the executor personality of the dataworker
recently, resulting in a significant number of OoG failures.
         
Side benefit: This might actually save a separate estimateGas() RPC 
call, so it's possibly faster and a bit cheaper in terms of RPC usage.

Fixes ACX-1236.

### CODE REVIEW

**Review by nicholaspai — APPROVED** [review]

File: src/clients/MultiCallerClient.ts [hunk 1]
```diff
@@ -380,9 +386,7 @@ export class MultiCallerClient {
         chunkSize
       );
       const multicallerTxnBundle = multicallerTxnChunks
-        .map((txnChunk) => {
-          return txnChunk.length > 1 ? this.buildMultiCallBundle(txnChunk) : txnChunk[0];
-        })
+        .map((txnChunk) => (txnChunk.length > 1 ? this.buildMultiCallBundle(txnChunk) : txnChunk[0]))
```
Comment (pxrl) [inline src/clients/MultiCallerClient.ts:389]: This is just a minor driveby refactor, since it can fit on a single line.

File: src/utils/TransactionUtils.ts [hunk 2]
```diff
@@ -12,7 +12,8 @@ dotenv.config();
 export type TransactionSimulationResult = {
   transaction: AugmentedTransaction;
   succeed: boolean;
-  reason: string;
+  gasLimit?: BigNumber;
+  reason?: string;
```
Comment (pxrl) [inline src/utils/TransactionUtils.ts:16]: Driveby change: made `reason` optional so we don't have to return it on simulation successes.

File: src/clients/MultiCallerClient.ts [hunk 3]
```diff
@@ -269,6 +272,7 @@ export class MultiCallerClient {
       contract: multisender,
       method: "aggregate",
       args: [callData],
+      gasLimitMultiplier: 1.2,
```
Comment (nicholaspai) [inline src/clients/MultiCallerClient.ts:275]: Nit can we make this 1.5 or even 2 to be safe?

**Review by nicholaspai — APPROVED** [review]
This is a good patch lgtm

File: src/clients/MultiCallerClient.ts [hunk 4]
```diff
@@ -188,8 +188,11 @@ export class MultiCallerClient {
       });
       if (!result.succeed) {
         batchesAllSucceeded = false;
+      } else {
+        batchTxns[idx].gasLimit = result.transaction.gasLimit;
```
Comment (james-a-morris) [inline src/clients/MultiCallerClient.ts:192]: Should the `batchesAllSucceeded` break the loop?

File: src/clients/MultiCallerClient.ts [hunk 5]
```diff
@@ -269,6 +272,7 @@ export class MultiCallerClient {
       contract: multisender,
       method: "aggregate",
       args: [callData],
+      gasLimitMultiplier: 2, // @note: Lots of headroom for the executor.
```
Comment (james-a-morris) [inline src/clients/MultiCallerClient.ts:275]: Should we proactively place this in a constant

File: src/clients/TransactionClient.ts [hunk 6]
```diff
@@ -24,6 +26,8 @@ export interface AugmentedTransaction {
   canFailInSimulation?: boolean;
 }
 
+const GAS_MULTIPLIER_1X = 1.0;
```
Comment (james-a-morris) [inline src/clients/TransactionClient.ts:29]: Should we place this in a constants file?

File: src/clients/TransactionClient.ts [hunk 7]
```diff
@@ -63,6 +67,16 @@ export class TransactionClient {
         this.logger.debug({ at: "TransactionClient#submit", message: `Using nonce ${nonce}.` });
       }
 
+      const gasLimitMultiplier = txn.gasLimitMultiplier ?? GAS_MULTIPLIER_1X;
+      if (gasLimitMultiplier != GAS_MULTIPLIER_1X) {
```
Comment (james-a-morris) [inline src/clients/TransactionClient.ts:71]: Since `gastLimitMultiplier` is potentially a non-integer (neither is `GAS_MULTIPLIER_1X`), should we perform fuzzy comparisons?

I.e. `Math.abs(gasLimitMultiplier - GAS_MULTIPLIER_1X) < .001`

**Review by james-a-morris — COMMENTED** [review]
Left a few nits

File: src/clients/MultiCallerClient.ts (hunk 5 above)
Comment (pxrl) [reply src/clients/MultiCallerClient.ts]: Addressed here: https://github.com/across-protocol/relayer-v2/pull/744/commits/b30cc5644a4ba131068af05cde6178caaa89ebdb

File: src/clients/TransactionClient.ts (hunk 7 above)
Comment (pxrl) [reply src/clients/TransactionClient.ts]: Fair call - have implemented something similar here: https://github.com/across-protocol/relayer-v2/pull/744/commits/6c42b6040d9c9d635a58375250ffaaf434719641

I don't think there's really a case for ever setting a multiplier < DEFAULT_GAS_MULTIPLIER, so I've just ignored that.

Comment (james-a-morris) [reply src/clients/TransactionClient.ts]: Fair enough!

**Review by james-a-morris — APPROVED** [review]
My comments have been addressed. Thanks!
